lib/spack/spack/package.py

```diff
@@ -1129,6 +1137,37 @@ def do_patch(self):
         else:
             touch(no_patches_file)
 
+    def patches_to_apply(self):
+        """If the patch set does not change between two invocations of spack,
+           then all patches in the set will be applied in the same order"""
+        patchesToApply = itertools.chain.from_iterable(
+            patch_list
+            for spec, patch_list in self.patches.items()
+            if self.spec.satisfies(spec))
+        return sorted(patchesToApply, key=lambda p: p.path_or_url)
```

Comment:
@adamjstewart This seems a likely candidate to inspect for the behavior seen in #7543 (we return the patches sorted by `url`, if I am not reading this wrongly)

Reply:
To add on these comments, I checked and we never really enforced a global order on patch application. What we were doing was to enforce a local order for the same constraint. 

Said otherwise, patches from the following directives:
```python
patch('patch1', when='constraint1')
patch('patch2', when='constraint2')
patch('patch3', when='constraint1')
```
are stored in packages like:
```python
{
  'constraint1': ['patch1', 'patch3'],
  'constraint2': ['patch2'],
}
```
using a dictionary. Before this PR we were granted that `patch3` would be applied after `patch1`, but we had no guarantee on the order of application for `patch2` with respect to the other two. After this PR we are granted that we apply patches sorted by their path or urls. **I think both strategies are wrong and we should apply patches in the order they are declared, if they match the constraint.**

I'll submit a minimal PR to recover the old behavior, and work on a more permanent fix later, if people agree with my analysis above. @tgamblin @mathstuf

Reply:
I agree with @alalazo.  @mathstuf is on vacation til fall so we'll have to go ahead without him.  Sorry I missed this in review!